src/Formats/IonWriter.cpp

```diff
@@ -0,0 +1,306 @@
+#include "IonWriter.h"
+
+#if USE_ION
+#include <decNumber/decContext.h>
+#include <Common/formatIPv6.h>
+
+namespace DB
+{
+
+IonWriter::IonWriter(WriteBuffer & out_, bool output_as_binary_) : out(out_), ion_buffer(out.available())
+{
+    // init output options
+    ION_WRITER_OPTIONS ion_options;
+    memset(&ion_options, 0, sizeof(ion_options));
+    ion_options.output_as_binary = output_as_binary_; // todo: Fix switch of writer type from format settings
+    if (!output_as_binary_)
+        ion_options.pretty_print = TRUE;
+
+    // Callback function called when flushing from an output stream
+    // flash is called only if we are at the zero nesting level, that is, not in a list or structure,
+    // which means one full ion message will be saved to the internal buffer,
+    // and then it will be transferred to WriteBuffer
+    ion_writer_callback = [this](_ion_user_stream * stream) -> iERR
+    {
+        // if limit is init it is call from flush
+        if (stream->limit == ion_buffer.data() + ion_buffer.size())
+        {
+            // Save current total written bytes
+            if ((stream->curr >= ion_buffer.data()) && (stream->curr <= ion_buffer.data() + ion_buffer.size()))
+                total_written_bytes = stream->curr - ion_buffer.data();
+
+            // Checking whether the buffer has overflowed; if so, expand and update the stream pointers
+            if (stream->curr == stream->limit)
+            {
+                // todo: Should we put a border on top of this?
+                ion_buffer.resize(ion_buffer.size() << 1);
+                stream->limit = ion_buffer.data() + ion_buffer.size();
+                stream->curr = ion_buffer.data() + total_written_bytes;
+            }
+        }
+        else
+        {
+            // else it is first call
+            total_written_bytes = 0;
+            stream->limit = ion_buffer.data() + ion_buffer.size();
+            stream->curr = ion_buffer.data();
+        }
+        return IERR_OK;
+    };
```

Comment:
So we write all our data into `ion_buffer`, resize it when needed, save total written bytes and write it to our `WriteBuffer` only in `IonWriter::final()`. It's not good, because we can output hundred gigabytes of data and all this data will be stored in memory in `ion_buffer` until we flush it in `IonWriter::final()` to our `WriteBuffer` (and it will happen only when all data is output).

Actually, I think here we don't need `ion_buffer` at all, we can use `WriteBuffer` directly. We can set `stream->curr` to `out.position()` and `stream->limit` to `out.buffer().end()`, so all data will be written directly to our `WiriteBuffer`. And during this callback when limit is reached we can set `out.position()` to `stream->curr` and call `out.next()` that will write all currently written data to underlying level (file/s3/nested buffer/etc) so we can reuse the buffer again and reset `stream->curr` to `out.position()` and `stream->limit` to `out.buffer().end()`.

But it will work only if ion library doesn't require all the data to be in continious range of memory  (because we will write the data to the same buffer and flush it over and over again).

BTW, when specifically this callback is called? Does it happen only on first init call and when `stream->curr == stream->limit` or it can happen also when `stream->curr < stream->limit`?

Reply:
This is still relevant